nixos/modules/services/hardware/kanata.nix

```diff
@@ -0,0 +1,173 @@
+        linux-dev ${keyboard.device})
+
+      ${keyboard.config}
+    '';
+
+  mkService = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata for ${keyboard.device}";
+
+      # Because path units are used to activate service units, which
+      # will start the old stopped services during "nixos-rebuild
+      # switch", stopIfChanged here is a workaround to make sure new
+      # services are running after "nixos-rebuild switch".
+      stopIfChanged = false;
+
+      serviceConfig = hardeningOptions // {
+        DynamicUser = true;
+        SupplementaryGroups = with config.users.groups; [
+          input.name
+          uinput.name
+        ];
+      };
+      script = ''
+        ${cfg.package}/bin/kanata \
+          --cfg ${mkConfig keyboard}
+      '';
+    };
+  };
+
+  mkPath = keyboard: {
+    name = mkName keyboard;
+    value = {
+      description = "kanata trigger for ${keyboard.device}";
+      wantedBy = [ "multi-user.target" ];
+      pathConfig = {
+        PathExists = keyboard.device;
+      };
+    };
+  };
+
+  hardeningOptions = {
+    DeviceAllow = [
+      "/dev/uinput w"
+      "char-input r"
+    ];
+    CapabilityBoundingSet = "";
+    DevicePolicy = "closed";
+    IPAddressDeny = "any";
+    LockPersonality = true;
+    MemoryDenyWriteExecute = true;
+    PrivateNetwork = true;
+    PrivateUsers = true;
+    ProcSubset = "pid";
+    ProtectClock = true;
+    ProtectControlGroups = true;
+    ProtectHome = true;
+    ProtectHostname = true;
+    ProtectKernelLogs = true;
+    ProtectKernelModules = true;
+    ProtectKernelTunables = true;
+    ProtectProc = "invisible";
+    RestrictAddressFamilies = "none";
+    RestrictNamespaces = true;
+    RestrictRealtime = true;
+    SystemCallArchitectures = "native";
+    SystemCallFilter = [
+      "@system-service"
+      "~@privileged"
+      "~@resources"
+    ];
+    UMask = "0077";
+  };
+in
+{
+  options.services.kanata = {
+    enable = mkEnableOption "kanata";
+    package = mkOption {
+      type = types.package;
+      default =
+        if any (keyboard: keyboard.enableCmd) (attrValues cfg.keyboards)
+        then pkgs.kanata-with-cmd
+        else pkgs.kanata;
+      defaultText = lib.literalExpression ''
+        if any (keyboard: keyboard.enableCmd) (attrValues cfg.keyboards)
+        then pkgs.kanata-with-cmd
+        else pkgs.kanata
+      '';
+      description = "kanata package to use.";
+    };
+    keyboards = mkOption {
+      type = types.attrsOf (types.submodule keyboard);
+      default = { };
+      description = "Keyboard configurations.";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    hardware.uinput.enable = true;
+
+    systemd = {
+      paths = trivial.pipe cfg.keyboards [
+        attrValues
+        (map mkPath)
+        listToAttrs
+      ];
+      services = trivial.pipe cfg.keyboards [
+        attrValues
+        (map mkService)
+        listToAttrs
+      ];
```

Comment:
these two could use `mapAttrs'` instead (which also removes the need for the `name` internal option on things)